File: library/src/main/java/me/shkschneider/skeleton/ui/AutoGridLayout.java
package me.shkschneider.skeleton.ui;

import android.annotation.TargetApi;
import android.content.Context;
import android.support.v4.view.GravityCompat;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.List;

import me.shkschneider.skeleton.helper.AndroidHelper;

// <https://github.com/AlbertGrobas/AutoLinearLayout>
public class AutoGridLayout extends FrameLayout {

    public static final int HORIZONTAL = 0;
    public static final int VERTICAL = 1;

    private int mOrientation;
    private int mGravity = (Gravity.TOP | Gravity.LEFT);

    private List<ViewPosition> mListPositions = new ArrayList<>();

    public AutoGridLayout(final Context context) {
        this(context, null);
    }

    public AutoGridLayout(final Context context, final AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public AutoGridLayout(final Context context, final AttributeSet attrs, final int defStyle) {
        super(context, attrs, defStyle);
        init(attrs);
    }

    @SuppressWarnings("unused")
    @TargetApi(AndroidHelper.API_21)
    public AutoGridLayout(final Context context, final AttributeSet attrs, final int defStyleAttr, final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(attrs);
    }

    private void init(final AttributeSet attrs) {
        final int orientation = ((attrs != null) ? attrs.getAttributeResourceValue(ViewHelper.ANDROIDXML, "orientation", -1) : -1);
        if (orientation == HORIZONTAL) {
            setHorizontalGravity(Gravity.CENTER_HORIZONTAL);
        }
        else if (orientation == VERTICAL) {
            setVerticalGravity(Gravity.CENTER_VERTICAL);
        }
        else {
            setGravity(Gravity.CENTER);
        }
    }

    @Override
    protected void onMeasure(final int widthMeasureSpec, final int heightMeasureSpec) {
        if (mOrientation == VERTICAL) {
            measureVertical(widthMeasureSpec, heightMeasureSpec);
        }
        else {
            measureHorizontal(widthMeasureSpec, heightMeasureSpec);
        }
    }

    private void measureHorizontal(final int widthMeasureSpec, final int heightMeasureSpec) {
        int wSize = MeasureSpec.getSize(widthMeasureSpec) - (getPaddingLeft() + getPaddingRight());
        if (MeasureSpec.getMode(widthMeasureSpec) == MeasureSpec.UNSPECIFIED) {
            wSize = Integer.MAX_VALUE;
        }
        final int count = getChildCount();
        int rowWidth = 0;
        int totalHeight = 0;
        int rowMaxHeight = 0;
        int childWidth;
        int maxRowHeight = getPaddingTop() + getPaddingBottom();
        for (int i = 0; i < count; i++) {
            final View child = getChildAt(i);
            if (child.getVisibility() != GONE) {
                measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0);
                final LayoutParams layoutParams = (LayoutParams) child.getLayoutParams();
                childWidth = child.getMeasuredWidth() + layoutParams.leftMargin + layoutParams.rightMargin;
                rowMaxHeight = Math.max(rowMaxHeight, child.getMeasuredHeight() + layoutParams.topMargin + layoutParams.bottomMargin);
                if ((childWidth + rowWidth) > wSize) {
                    totalHeight += rowMaxHeight;
                    maxRowHeight = Math.max(maxRowHeight, rowWidth);
                    rowWidth = childWidth;
                    rowMaxHeight = 0;
                }
                else {
                    rowWidth += childWidth;
                }
            }
        }
        if (rowWidth != 0) {
            maxRowHeight = Math.max(maxRowHeight, rowWidth);
            totalHeight += rowMaxHeight;
        }
        if (MeasureSpec.getMode(widthMeasureSpec) == MeasureSpec.UNSPECIFIED) {
            wSize = maxRowHeight - (getPaddingLeft() + getPaddingRight());
        }
        setMeasuredDimension(resolveSize(wSize, widthMeasureSpec), resolveSize(totalHeight + getPaddingTop() + getPaddingBottom(), heightMeasureSpec));
    }

    private void measureVertical(final int widthMeasureSpec, final int heightMeasureSpec) {
        int hSize = MeasureSpec.getSize(heightMeasureSpec) - (getPaddingTop() + getPaddingBottom());
        final int count = getChildCount();
        int columnHeight = 0;
        int totalWidth = 0, maxColumnHeight = 0;
        int columnMaxWidth = 0;
        int childHeight;
        if (MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.UNSPECIFIED) {
            hSize = Integer.MAX_VALUE;
        }
        for (int i = 0; i < count; i++) {
            final View child = getChildAt(i);
            if (child.getVisibility() != GONE) {
                measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0);
                final LayoutParams layoutParams = (LayoutParams) child.getLayoutParams();
                childHeight = child.getMeasuredHeight() + layoutParams.topMargin + layoutParams.bottomMargin;
                columnMaxWidth = Math.max(columnMaxWidth, child.getMeasuredWidth() + layoutParams.leftMargin + layoutParams.rightMargin);
                if ((childHeight + columnHeight) > hSize) {
                    totalWidth += columnMaxWidth;
                    maxColumnHeight = Math.max(maxColumnHeight, columnHeight);
                    columnHeight = childHeight;
                    columnMaxWidth = 0;
                }
                else {
                    columnHeight += childHeight;
                }
            }
        }
        if (columnHeight != 0) {
            maxColumnHeight = Math.max(maxColumnHeight, columnHeight);
            totalWidth += columnMaxWidth;
        }
        if (MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.UNSPECIFIED) {
            hSize = maxColumnHeight - (getPaddingTop() + getPaddingBottom());
        }
        setMeasuredDimension(resolveSize(totalWidth + getPaddingRight() + getPaddingLeft(), widthMeasureSpec), resolveSize(hSize, heightMeasureSpec));
    }

    @Override
    protected void onLayout(final boolean changed, final int left, final int top, final int right, final int bottom) {
        mListPositions.clear();
        if (mOrientation == VERTICAL) {
            layoutVertical(left, top, right, bottom);
        }
        else {
            layoutHorizontal(left, top, right, bottom);
        }
    }

    private void layoutVertical(final int left, final int top, final int right, final int bottom) {
        final int count = getChildCount();
        if (count == 0) {
            return;
        }
        final int width = right - getPaddingLeft() - left - getPaddingRight();
        final int height = bottom - getPaddingTop() - top - getPaddingBottom();
        int childTop = getPaddingTop();
        int childLeft = getPaddingLeft();
        int totalHorizontal = getPaddingLeft() + getPaddingRight();
        int totalVertical = 0;
        int column = 0;
        int maxChildWidth = 0;
        for (int i = 0; i < count; i++) {
            final View child = getChildAt(i);
            if ((child != null) && (child.getVisibility() != View.GONE)) {
                if ((child.getMeasuredHeight() == 0) || (child.getMeasuredWidth() == 0)) {
                    child.measure(MeasureSpec.makeMeasureSpec(width, MeasureSpec.AT_MOST), MeasureSpec.makeMeasureSpec(height, MeasureSpec.AT_MOST));
                }
                final LayoutParams layoutParams = (LayoutParams) child.getLayoutParams();
                final int childWidth = child.getMeasuredWidth();
                final int childHeight = child.getMeasuredHeight();
                if ((childTop + childHeight + layoutParams.topMargin + layoutParams.bottomMargin) > (height + getPaddingTop())) {
                    updateChildPositionVertical(height, totalVertical, column, maxChildWidth);
                    childTop = getPaddingTop();
                    childLeft += maxChildWidth;
                    maxChildWidth = 0;
                    column++;
                    totalVertical = 0;
                }
                childTop += layoutParams.topMargin;
                mListPositions.add(new ViewPosition(childLeft, childTop, column));
                final int currentWidth = childWidth + layoutParams.leftMargin + layoutParams.rightMargin;
                if (maxChildWidth < currentWidth) {
                    maxChildWidth = currentWidth;
                }
                childTop += childHeight + layoutParams.bottomMargin;
                totalVertical += childHeight + layoutParams.topMargin + layoutParams.bottomMargin;
            }
        }
        updateChildPositionVertical(height, totalVertical, column, maxChildWidth);
        totalHorizontal += (childLeft + maxChildWidth);
        updateChildPositionHorizontal(width, totalHorizontal, column, 0);
        // mListPositions.clear();
    }

    private void layoutHorizontal(final int left, final int top, final int right, final int bottom) {
        final int count = getChildCount();
        if (count == 0) {
            return;
        }
        final int width = right - getPaddingLeft() - left - getPaddingRight();
        final int height = bottom - getPaddingTop() - top - getPaddingBottom();
        int childTop = getPaddingTop();
        int childLeft = getPaddingLeft();
        int totalHorizontal = 0;
        int totalVertical = getPaddingTop() + getPaddingBottom();
        int row = 0;
        int maxChildHeight = 0;
        for (int i = 0; i < count; i++) {
            final View child = getChildAt(i);
            if ((child != null) && (child.getVisibility() != View.GONE)) {
                if ((child.getMeasuredHeight() == 0) || (child.getMeasuredWidth() == 0)) {
                    child.measure(MeasureSpec.makeMeasureSpec(width, MeasureSpec.AT_MOST), MeasureSpec.makeMeasureSpec(height, MeasureSpec.AT_MOST));
                }
                final LayoutParams layoutParams = (LayoutParams) child.getLayoutParams();
                final int childWidth = child.getMeasuredWidth();
                final int childHeight = child.getMeasuredHeight();
                if ((childLeft + childWidth + layoutParams.leftMargin + layoutParams.rightMargin) > (width + getPaddingLeft())) {
                    updateChildPositionHorizontal(width, totalHorizontal, row, maxChildHeight);
                    childLeft = getPaddingLeft();
                    childTop += maxChildHeight;
                    maxChildHeight = 0;
                    row++;
                    totalHorizontal = 0;
                }
                childLeft += layoutParams.leftMargin;
                mListPositions.add(new ViewPosition(childLeft, childTop, row));
                int currentHeight = childHeight + layoutParams.topMargin + layoutParams.bottomMargin;
                if (maxChildHeight < currentHeight) {
                    maxChildHeight = currentHeight;
                }
                childLeft += (childWidth + layoutParams.rightMargin);
                totalHorizontal += (childWidth + layoutParams.rightMargin + layoutParams.leftMargin);
            }
        }
        updateChildPositionHorizontal(width, totalHorizontal, row, maxChildHeight);
        totalVertical += (childTop + maxChildHeight);
        updateChildPositionVertical(height, totalVertical, row, 0);
        // mListPositions.clear();
    }

    private void updateChildPositionVertical(final int height, final int totalSize, final int column, final int maxChildWidth) {
        for (int i = 0; i < mListPositions.size(); i++) {
            final ViewPosition viewPosition = mListPositions.get(i);
            final View child = getChildAt(i);
            if ((mOrientation == HORIZONTAL) || (viewPosition.position == column)) {
                updateTopPositionByGravity(viewPosition, height - totalSize, mGravity);
            }
            if ((mOrientation == VERTICAL) && (viewPosition.position == column)) {
                final LayoutParams layoutParams = (LayoutParams) child.getLayoutParams();
                final int size = maxChildWidth - child.getMeasuredWidth() - layoutParams.leftMargin - layoutParams.rightMargin;
                updateLeftPositionByGravity(viewPosition, size, layoutParams.gravity);
            }
            if (mOrientation == HORIZONTAL) {
                layout(child, viewPosition);
            }
        }
    }

    private void updateChildPositionHorizontal(final int width, final int totalSize, final int row, final int maxChildHeight) {
        for (int i = 0; i < mListPositions.size(); i++) {
            final ViewPosition viewPosition = mListPositions.get(i);
            final View child = getChildAt(i);
            if ((mOrientation == VERTICAL) || (viewPosition.position == row)) {
                updateLeftPositionByGravity(viewPosition, width - totalSize, mGravity);
            }
            if ((mOrientation == HORIZONTAL) && (viewPosition.position == row)) {
                final LayoutParams layoutParams = (LayoutParams) child.getLayoutParams();
                final int size = maxChildHeight - child.getMeasuredHeight() - layoutParams.topMargin - layoutParams.bottomMargin;
                updateTopPositionByGravity(viewPosition, size, layoutParams.gravity);
            }
            if (mOrientation == VERTICAL) {
                layout(child, viewPosition);
            }
        }
    }

    private void updateLeftPositionByGravity(final ViewPosition viewPosition, final int size, final int gravity) {
        switch ((gravity & Gravity.HORIZONTAL_GRAVITY_MASK)) {
            case Gravity.RIGHT:
                viewPosition.left += ((size > 0) ? size : 0);
                break ;
            case Gravity.CENTER_HORIZONTAL:
                viewPosition.left += (((size > 0) ? size : 0) / 2);
                break ;
        }
    }

    private void updateTopPositionByGravity(final ViewPosition viewPosition, final int size, final int gravity) {
        switch ((gravity & Gravity.VERTICAL_GRAVITY_MASK)) {
            case Gravity.BOTTOM:
                viewPosition.top += ((size > 0) ? size : 0);
                break ;
            case Gravity.CENTER_VERTICAL:
                viewPosition.top += (((size > 0) ? size : 0) / 2);
                break ;
        }
    }

    private void layout(final View child, final ViewPosition viewPosition) {
        final LayoutParams layoutParams = (LayoutParams) child.getLayoutParams();
        if (mOrientation == HORIZONTAL) {
            child.layout(viewPosition.left,
                    viewPosition.top + layoutParams.topMargin,
                    viewPosition.left + child.getMeasuredWidth(),
                    viewPosition.top + child.getMeasuredHeight() + layoutParams.topMargin);
        }
        else {
            child.layout(viewPosition.left + layoutParams.leftMargin,
                    viewPosition.top,
                    viewPosition.left + child.getMeasuredWidth() + layoutParams.leftMargin,
                    viewPosition.top + child.getMeasuredHeight());
        }
    }

    public void setGravity(int gravity) {
        if (mGravity != gravity) {
            if ((gravity & Gravity.RELATIVE_HORIZONTAL_GRAVITY_MASK) == 0) {
                gravity |= GravityCompat.START;
            }
            if ((gravity & Gravity.VERTICAL_GRAVITY_MASK) == 0) {
                gravity |= Gravity.TOP;
            }
            mGravity = gravity;
            requestLayout();
        }
    }

    public void setHorizontalGravity(final int horizontalGravity) {
        final int gravity = (horizontalGravity & GravityCompat.RELATIVE_HORIZONTAL_GRAVITY_MASK);
        if ((mGravity & Gravity.RELATIVE_HORIZONTAL_GRAVITY_MASK) != gravity) {
            mGravity = ((mGravity & ~GravityCompat.RELATIVE_HORIZONTAL_GRAVITY_MASK) | gravity);
            requestLayout();
        }
    }

    public void setVerticalGravity(final int verticalGravity) {
        final int gravity = (verticalGravity & Gravity.VERTICAL_GRAVITY_MASK);
        if ((mGravity & Gravity.VERTICAL_GRAVITY_MASK) != gravity) {
            mGravity = ((mGravity & ~Gravity.VERTICAL_GRAVITY_MASK) | gravity);
            requestLayout();
        }
    }

    public void setOrientation(final int orientation) {
        if ((orientation != HORIZONTAL) && (orientation != VERTICAL)) {
            return;
        }
        if (mOrientation != orientation) {
            mOrientation = orientation;
            requestLayout();
        }
    }

    public int getOrientation() {
        return mOrientation;
    }

    static class ViewPosition {

        int left;
        int top;
        int position; //row or column

        public ViewPosition(final int left, final int top, final int position) {
            this.left = left;
            this.top = top;
            this.position = position;
        }

        @Override
        public String toString() {
            return "left:" + left + "|top:" + top + "|position:" + position;
        }

    }

}

File: core/src/main/kotlin/me/shkschneider/skeleton/ui/widget/AutoGridLayout.kt
package me.shkschneider.skeleton.ui.widget

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import androidx.core.view.GravityCompat
import androidx.recyclerview.widget.RecyclerView
import me.shkschneider.skeleton.extensions.android.ViewHelper
import me.shkschneider.skeleton.extensions.has
import me.shkschneider.skeleton.helper.AndroidHelper

private const val HORIZONTAL = LinearLayout.HORIZONTAL
private const val VERTICAL = LinearLayout.VERTICAL

// <https://github.com/AlbertGrobas/AutoLinearLayout>
class AutoGridLayout : FrameLayout {

    private var orientation = 0
        set(orientation) {
            if (orientation != HORIZONTAL && orientation != VERTICAL) {
                return
            }
            if (this.orientation != orientation) {
                field = orientation
                requestLayout()
            }
        }
    private var gravity = Gravity.TOP or Gravity.START
    private val positions = ArrayList<ViewPosition>()

    constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : super(context, attrs, defStyleAttr) {
        init(attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(attrs)
    }

    private fun init(attrs: AttributeSet?) {
        val orientation = attrs?.getAttributeResourceValue(ViewHelper.ANDROIDXML, "orientation", -1)
                ?: -1
        when (orientation) {
            HORIZONTAL -> setHorizontalGravity(Gravity.CENTER_HORIZONTAL)
            VERTICAL -> setVerticalGravity(Gravity.CENTER_VERTICAL)
            else -> setGravity(Gravity.CENTER)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (orientation == VERTICAL) {
            measureVertical(widthMeasureSpec, heightMeasureSpec)
        } else {
            measureHorizontal(widthMeasureSpec, heightMeasureSpec)
        }
    }

    private fun measureHorizontal(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var wSize = View.MeasureSpec.getSize(widthMeasureSpec) - (paddingLeft + paddingRight)
        if (View.MeasureSpec.getMode(widthMeasureSpec) == View.MeasureSpec.UNSPECIFIED) {
            wSize = Integer.MAX_VALUE
        }
        val count = childCount
        var rowWidth = 0
        var totalHeight = 0
        var rowMaxHeight = 0
        var childWidth: Int
        var maxRowHeight = paddingTop + paddingBottom
        for (i in 0 until count) {
            val child = getChildAt(i)
            if (child.visibility != View.GONE) {
                measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0)
                val layoutParams = child.layoutParams as FrameLayout.LayoutParams
                childWidth = child.measuredWidth + layoutParams.leftMargin + layoutParams.rightMargin
                rowMaxHeight = Math.max(rowMaxHeight, child.measuredHeight + layoutParams.topMargin + layoutParams.bottomMargin)
                if (childWidth + rowWidth > wSize) {
                    totalHeight += rowMaxHeight
                    maxRowHeight = Math.max(maxRowHeight, rowWidth)
                    rowWidth = childWidth
                    rowMaxHeight = 0
                } else {
                    rowWidth += childWidth
                }
            }
        }
        if (rowWidth != 0) {
            maxRowHeight = Math.max(maxRowHeight, rowWidth)
            totalHeight += rowMaxHeight
        }
        if (View.MeasureSpec.getMode(widthMeasureSpec) == View.MeasureSpec.UNSPECIFIED) {
            wSize = maxRowHeight - (paddingLeft + paddingRight)
        }
        setMeasuredDimension(View.resolveSize(wSize, widthMeasureSpec), View.resolveSize(totalHeight + paddingTop + paddingBottom, heightMeasureSpec))
    }

    private fun measureVertical(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var hSize = View.MeasureSpec.getSize(heightMeasureSpec) - (paddingTop + paddingBottom)
        val count = childCount
        var columnHeight = 0
        var totalWidth = 0
        var maxColumnHeight = 0
        var columnMaxWidth = 0
        var childHeight: Int
        if (View.MeasureSpec.getMode(heightMeasureSpec) == View.MeasureSpec.UNSPECIFIED) {
            hSize = Integer.MAX_VALUE
        }
        for (i in 0 until count) {
            val child = getChildAt(i)
            if (child.visibility != View.GONE) {
                measureChildWithMargins(child, widthMeasureSpec, 0, heightMeasureSpec, 0)
                val layoutParams = child.layoutParams as FrameLayout.LayoutParams
                childHeight = child.measuredHeight + layoutParams.topMargin + layoutParams.bottomMargin
                columnMaxWidth = Math.max(columnMaxWidth, child.measuredWidth + layoutParams.leftMargin + layoutParams.rightMargin)
                if (childHeight + columnHeight > hSize) {
                    totalWidth += columnMaxWidth
                    maxColumnHeight = Math.max(maxColumnHeight, columnHeight)
                    columnHeight = childHeight
                    columnMaxWidth = 0
                } else {
                    columnHeight += childHeight
                }
            }
        }
        if (columnHeight != 0) {
            maxColumnHeight = Math.max(maxColumnHeight, columnHeight)
            totalWidth += columnMaxWidth
        }
        if (View.MeasureSpec.getMode(heightMeasureSpec) == View.MeasureSpec.UNSPECIFIED) {
            hSize = maxColumnHeight - (paddingTop + paddingBottom)
        }
        setMeasuredDimension(View.resolveSize(totalWidth + paddingRight + paddingLeft, widthMeasureSpec), View.resolveSize(hSize, heightMeasureSpec))
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        positions.clear()
        if (orientation == VERTICAL) {
            layoutVertical(left, top, right, bottom)
        } else {
            layoutHorizontal(left, top, right, bottom)
        }
    }

    private fun layoutVertical(left: Int, top: Int, right: Int, bottom: Int) {
        if (childCount == 0) {
            return
        }
        val width = right - paddingLeft - left - paddingRight
        val height = bottom - paddingTop - top - paddingBottom
        var childTop = paddingTop
        var childLeft = paddingLeft
        var totalHorizontal = paddingLeft + paddingRight
        var totalVertical = 0
        var column = 0
        var maxChildWidth = 0
        for (i in 0 until childCount) {
            val viewGroup = getChildAt(i) ?: continue
            if (viewGroup.visibility != View.GONE) {
                if (viewGroup.measuredHeight == 0 || viewGroup.measuredWidth == 0) {
                    viewGroup.measure(View.MeasureSpec.makeMeasureSpec(width, View.MeasureSpec.AT_MOST), View.MeasureSpec.makeMeasureSpec(height, View.MeasureSpec.AT_MOST))
                }
                val layoutParams = viewGroup.layoutParams as FrameLayout.LayoutParams
                val childWidth = viewGroup.measuredWidth
                val childHeight = viewGroup.measuredHeight
                if (childTop + childHeight + layoutParams.topMargin + layoutParams.bottomMargin > height + paddingTop) {
                    updateChildPositionVertical(height, totalVertical, column, maxChildWidth)
                    childTop = paddingTop
                    childLeft += maxChildWidth
                    maxChildWidth = 0
                    column++
                    totalVertical = 0
                }
                childTop += layoutParams.topMargin
                positions.add(ViewPosition(childLeft, childTop, column))
                val currentWidth = childWidth + layoutParams.leftMargin + layoutParams.rightMargin
                if (maxChildWidth < currentWidth) {
                    maxChildWidth = currentWidth
                }
                childTop += childHeight + layoutParams.bottomMargin
                totalVertical += childHeight + layoutParams.topMargin + layoutParams.bottomMargin
            }
        }
        updateChildPositionVertical(height, totalVertical, column, maxChildWidth)
        totalHorizontal += childLeft + maxChildWidth
        updateChildPositionHorizontal(width, totalHorizontal, column, 0)
        // positions.clear();
    }

    private fun layoutHorizontal(left: Int, top: Int, right: Int, bottom: Int) {
        val count = childCount
        if (count == 0) {
            return
        }
        val width = right - paddingLeft - left - paddingRight
        val height = bottom - paddingTop - top - paddingBottom
        var childTop = paddingTop
        var childLeft = paddingLeft
        var totalHorizontal = 0
        var totalVertical = paddingTop + paddingBottom
        var row = 0
        var maxChildHeight = 0
        for (i in 0 until count) {
            val viewGroup = getChildAt(i) ?: continue
            if (viewGroup.visibility != View.GONE) {
                if (viewGroup.measuredHeight == 0 || viewGroup.measuredWidth == 0) {
                    viewGroup.measure(View.MeasureSpec.makeMeasureSpec(width, View.MeasureSpec.AT_MOST), View.MeasureSpec.makeMeasureSpec(height, View.MeasureSpec.AT_MOST))
                }
                val layoutParams = viewGroup.layoutParams as FrameLayout.LayoutParams
                val childWidth = viewGroup.measuredWidth
                val childHeight = viewGroup.measuredHeight
                if (childLeft + childWidth + layoutParams.leftMargin + layoutParams.rightMargin > width + paddingLeft) {
                    updateChildPositionHorizontal(width, totalHorizontal, row, maxChildHeight)
                    childLeft = paddingLeft
                    childTop += maxChildHeight
                    maxChildHeight = 0
                    row++
                    totalHorizontal = 0
                }
                childLeft += layoutParams.leftMargin
                positions.add(ViewPosition(childLeft, childTop, row))
                val currentHeight = childHeight + layoutParams.topMargin + layoutParams.bottomMargin
                if (maxChildHeight < currentHeight) {
                    maxChildHeight = currentHeight
                }
                childLeft += childWidth + layoutParams.rightMargin
                totalHorizontal += childWidth + layoutParams.rightMargin + layoutParams.leftMargin
            }
        }
        updateChildPositionHorizontal(width, totalHorizontal, row, maxChildHeight)
        totalVertical += childTop + maxChildHeight
        updateChildPositionVertical(height, totalVertical, row, 0)
        // positions.clear();
    }

    private fun updateChildPositionVertical(height: Int, totalSize: Int, column: Int, maxChildWidth: Int) {
        for (i in positions.indices) {
            val viewPosition = positions[i]
            val child = getChildAt(i)
            if (orientation == HORIZONTAL || viewPosition.position == column) {
                updateTopPositionByGravity(viewPosition, height - totalSize, gravity)
            }
            if (orientation == VERTICAL && viewPosition.position == column) {
                val layoutParams = child.layoutParams as FrameLayout.LayoutParams
                val size = maxChildWidth - child.measuredWidth - layoutParams.leftMargin - layoutParams.rightMargin
                updateLeftPositionByGravity(viewPosition, size, layoutParams.gravity)
            }
            if (orientation == HORIZONTAL) {
                layout(child, viewPosition)
            }
        }
    }

    private fun updateChildPositionHorizontal(width: Int, totalSize: Int, row: Int, maxChildHeight: Int) {
        for (i in positions.indices) {
            val viewPosition = positions[i]
            val child = getChildAt(i)
            if (orientation == VERTICAL || viewPosition.position == row) {
                updateLeftPositionByGravity(viewPosition, width - totalSize, gravity)
            }
            if (orientation == HORIZONTAL && viewPosition.position == row) {
                val layoutParams = child.layoutParams as FrameLayout.LayoutParams
                val size = maxChildHeight - child.measuredHeight - layoutParams.topMargin - layoutParams.bottomMargin
                updateTopPositionByGravity(viewPosition, size, layoutParams.gravity)
            }
            if (orientation == VERTICAL) {
                layout(child, viewPosition)
            }
        }
    }

    private fun updateLeftPositionByGravity(viewPosition: ViewPosition, size: Int, gravity: Int) {
        when (gravity and Gravity.HORIZONTAL_GRAVITY_MASK) {
            Gravity.END -> viewPosition.left += if (size > 0) size else 0
            Gravity.CENTER_HORIZONTAL -> viewPosition.left += (if (size > 0) size else 0) / 2
        }
    }

    private fun updateTopPositionByGravity(viewPosition: ViewPosition, size: Int, gravity: Int) {
        when (gravity and Gravity.VERTICAL_GRAVITY_MASK) {
            Gravity.BOTTOM -> viewPosition.top += if (size > 0) size else 0
            Gravity.CENTER_VERTICAL -> viewPosition.top += (if (size > 0) size else 0) / 2
        }
    }

    private fun layout(child: View, viewPosition: ViewPosition) {
        val layoutParams = child.layoutParams as FrameLayout.LayoutParams
        if (orientation == HORIZONTAL) {
            child.layout(viewPosition.left,
                    viewPosition.top + layoutParams.topMargin,
                    viewPosition.left + child.measuredWidth,
                    viewPosition.top + child.measuredHeight + layoutParams.topMargin)
        } else {
            child.layout(viewPosition.left + layoutParams.leftMargin,
                    viewPosition.top,
                    viewPosition.left + child.measuredWidth + layoutParams.leftMargin,
                    viewPosition.top + child.measuredHeight)
        }
    }

    fun setGravity(gravity: Int) {
        if (this.gravity != gravity) {
            if (gravity.has(Gravity.RELATIVE_HORIZONTAL_GRAVITY_MASK)) {
                this.gravity = gravity or GravityCompat.START
            }
            if (gravity.has(Gravity.VERTICAL_GRAVITY_MASK)) {
                this.gravity = gravity or Gravity.TOP
            }
            requestLayout()
        }
    }

    fun setHorizontalGravity(horizontalGravity: Int) {
        val gravity = horizontalGravity and GravityCompat.RELATIVE_HORIZONTAL_GRAVITY_MASK
        if (this.gravity.has(Gravity.RELATIVE_HORIZONTAL_GRAVITY_MASK)) {
            this.gravity = this.gravity and GravityCompat.RELATIVE_HORIZONTAL_GRAVITY_MASK.inv() or gravity
            requestLayout()
        }
    }

    fun setVerticalGravity(verticalGravity: Int) {
        val gravity = verticalGravity and Gravity.VERTICAL_GRAVITY_MASK
        if (this.gravity.has(Gravity.VERTICAL_GRAVITY_MASK)) {
            this.gravity = this.gravity and Gravity.VERTICAL_GRAVITY_MASK.inv() or gravity
            requestLayout()
        }
    }

    internal class ViewPosition(var left: Int, var top: Int, val position: Int) {

        override fun toString(): String {
            return "left:$left|top:$top|position:$position"
        }

    }

}
